### DISCUSSION TITLE
Feature Request: Live Resolution Change

### DISCUSSION THREAD

**Discussion by jsn081** [GerryDazoo/Slinger#77] [Ideas]
Awesome job.  This works well.  I have it running on Ubuntu and streaming via VLC to Windows.

One feature request: Can we set four resolutions (basic, low, high, best) with pre-defined resolutions and bitrates, and have a toggle (radio buttons or drop-down) on the remote page to switch between them?  It would make a nice user-facing way to dynamically get around lag, instead of having to edit an INI.  I think the hardware allows it, because the app handled the downgrades/upgrades dynamically and automatically.

**Comment by GerryDazoo** [comment]
What you're asking for is a lot of work and probably won't happen (at least not by me), because everybody's definition if good, better, best will be different. So you're going to be stuck editing your own config.ini files to suit your specific hardware, bandwidth and viewing preferences.
So may I suggest make 4 config files with your own good better best definitions and then make some VLC shortcuts for each one. Then you can just click whichever VLC you want.

PS. the next release will make this easier by allowing you to put this all in one config.ini file.

**Reply by jsn081** [reply]
The thought was to put the settings IN the .ini file - something you pre-configure.  So for example you have variables for:

res_basic = 5
bw_basic = 200
res_low = 5
bw_low = 500
res_high = 12
bw_high = 1000

...etc.  So you pre-define them to your use case, and when you hit the toggle on the remote page, it sends a post command just like the buttons do - but it sends the byte sequence to the slingbox to change the bw/resolution.

Does that make sense?  You've already got all of the parts working (a cgi-post, the byte sequence to set resolution, etc), it would just be a minor change to the remote page to add the toggle, the INI file to add the new configurable values, and the python file to accept the new post value and send the corresponding resolution and bandwidth and resolution command to the slingbox.

**Comment by GerryDazoo** [comment]
Bu if you have to come up with the settings yourself why don't you just cut and paste them into your config.ini. Sounds like you want me to do hours of work to save you a few minutes. Sorry, not going to happen.

**Reply by jsn081** [reply]
No, the idea was to be able to change the resolution on the fly from the stream's endpoint if the data starts to lag.  My mother slings her tivo to a cabin, but the sling app had resolution settings so if the stream started falling behind because of network lag, she could downgrade and keep watching.

The issues with the current method are 1) She will never know how to edit an INI file, and 2) When she's up north and not at home (where the server is), she can't log into the box to change any settings.  But even people like myself would benefit - selecting a drop down is far easier than VNCing into my house and logging into Debian just to change the bitrate.

If you don't want to do it that's fine, but there's a huge usability reason to do it - and even the old app did it (they didn't make those rates and resolutions customizable, they pre-selected the 4 settings).

I just entered my busy season, otherwise I'd tear into the code and do it myself.  If someone else hasn't done it by year's end, I'll take a stab at it.

**Reply by GerryDazoo** [reply]
I think you're missing something important, you can have more than one config.ini. If you want to have different streaming parameters make two config.ini file with different names pointing to the same slingbox and pass them into the server on the command line.
Then just point your favorite media player at the one you want. I do this on a daily basis. I have two configs one for streaming HD to my big screen at home and one for when I want to watch sometime on my phone while using mobile data. The next release will allow you to put all this into one config file and let you select the profile by name   i.e.  http://my_place:port/500/HD or http://my_place:port/500/Mobile.

**Reply by GerryDazoo** [reply]
But for the next week or two you would access the two different profiles using different port numbers
http://my_place:HD_port/slingbox  or http://my_place:Mobile_port/slingbox

**Comment by rbf121** [comment]
Another option would be to get the best possible quality from your slingbox and then put that stream into something else like plex/emby to transcode when needed.  I have this running in plex and can just let plex do the transcoding when needed, if your server can handle transcoding.  I just tested it and my server did hw transcoding to change the bitrate/resolution.

**Comment by Pthepeg** [comment]
I liked the idea, and I've put together a web interface that lets you pick from 3 resolutions. It launches slingserver with the corresponding config.ini (there are 3 config.ini files, one for each resolution)
The player has to be relaunched when switching resolution but the stream address stays the same.

**Reply by GerryDazoo** [reply]
You don't need 3 config.ini's. Please check out the recent release notes on the unified_config.ini support. One file as many slingboxes as you want.

**Reply by Pthepeg** [reply]
A single config.ini would put each resolution on a different port, right?

**Reply by GerryDazoo** [reply]
No, one server port. Here's a sample for something like what you want to do,

[SLINGBOXES]
sb1=500
sb2=Solo
sb3=M1
sb4=500lbr
sb5=500audio

[500]
sbtype="350/500/M1"
password=3s4z2141a121XXX
ipaddress=192.168.117.196
port=5301
Resolution=12
FrameRate=60
VideoBandwidth=2000
VideoSmoothness=63
VideoSource=1
maxstreams=3
;StartChannel=155
;RemoteLock=yes
Remote=500_remote.txt

[Solo]
sbtype="Solo/Pro/ProHD"
password=Toast     
ipaddress=192.168.117.101
port=5001
Resolution=5
FrameRate=30
VideoBandwidth=2000
VideoSmoothness=63
VideoSource=2
RemoteLock=yes
;StartChannel=155
Remote=solo_remote.txt

[M1]
sbtype="350/500/M1"
password=3s4z2141a1XXX    
ipaddress=192.168.117.129
port=5301
Resolution=12
FrameRate=30
VideoBandwidth=2000
VideoSmoothness=63
VideoSource=1
;Remote=m1_remote.txt
Remote=skyhdRemote.html


[500lbr]
sbtype="350/500/M1"
password=3s4z2141a121xxx
ipaddress=192.168.117.196
port=5301
Resolution=5
FrameRate=30
;Min 50 Max 8000
VideoBandwidth=1000
VideoSource=1
;VideoSmoothness=63
;IframeRate=5
AudioBitRate=96

[500audio]
sbtype="350/500/M1"
password=3s4z2141a121xxx
ipaddress=192.168.117.196
port=5301

; Following are the default values
; Default resolution when server starts
Resolution=0
AudioBitRate=96

[SERVER]
; local port number for the server to listen on for connections
; port number+1 is used by the remote control code.
port=65432
maxremotestreams=3
URLbase=Gerry

**Reply by Pthepeg** [reply]
Thanks but that's not what I want to do. The discussion is about different resolutions on a single slingbox. So is it possible to define the same slingbox 3 times in the ini file, 3 different resolutions?

**Reply by GerryDazoo** [reply]
Yes. In the example 500 and 500lbr and 500audio are three different resolutions for the same slingbox.  So yes I think it is what you want to do. Sorry I stuck in the Solo and M1 stuff. I just copied my current file.

**Reply by Pthepeg** [reply]
Ok, I missed that detail, that you can have the same slingbox defined multiple times in one config file. So from your example, from one slingbox, three resolutions are available at the player: 

http://serverip:serverport/Gerry/500   #resolution 12
http://serverip:serverport/Gerry/500lbr  #resolution 5
http://serverip:serverport/Gerry/500audio    #resolution 0

For this example, it might make sense to name the slingboxes by their resolution so its more obvious reading the URL. Seems to be a good feature, simplifies port forwarding configuration on the router, easy enough to pick your resolution in the player, and likely everyone is satisfied with that.
What I was after is a single URL at the player. No matter what resolution, the player would always connect to the same URL. So I have that by starting slingserver with one of a set of  different config.ini files, keeping the stream URL the same but changing the resolution setting from one config to the next. (I still have to use multiple different config files to do it, but that doesn't matter to me, once its setup I can forget about it.) Maybe I'm the only one that likes the feature of a single URL at the player, and TBH it may not justify the effort to create the web pages etc to do it. But I'm having fun. Thanks.

**Reply by GerryDazoo** [reply]
Sure, you can call them whatever you want. 
If you don't want to change the URL you'll have to swap out the config.ini files. So on the next stream start it will re-read the video parameters from the new file. You don't have to restart the server.

**Reply by jsn081** [reply]
I want to look into how the Slingbox app allowed the bitrate to be changed mid-stream.  It tells me that the Slingbox is capable of receiving the setting changes live, but I'm not sure if VLC will be happy if it sees a sudden bitrate change.  I'm still in work hell, I hope to be able to play around with this in mid-December if someone else doesn't beat me to it.  My end-goal is a toggle box on the remote that live-changes the resolution on the open stream.  The pipe dream (if I can get that to work) is code that watches the dropped frames counter on VLC and makes the change automatically.

**Reply by GerryDazoo** [reply]
Yes, VLC is not happy doing this. Here's some notes for the upcoming release...

The issue is (was) when changing channels the SB restarts the stream with the new metadata embedded in the packets. The players weren’t expecting that and would freeze or drop the connection. So now, on the channel change Slinger drops the connection to force the clients to reconnect with the new stream parameters. 
On VLC turn on the Loop ALL/Loop 1 to get it to reconnect when the stream drops. I know it’s not pretty and takes a few seconds to resync but at least it works.

In your case this would apply to a dynamic video format change mid-stream.

**Reply by mdehrlich** [reply]
I vote for a special symbol in the channel input to redefine resolution.  Much like the test keycode.  Something like !16 sets to resolution 16.

**Reply by GerryDazoo** [reply]
I used to support something like that but I took it out due to all the issues with the media players not being able to handle the dynamic format change. But also just changing resolution probably isn't sufficient. In most cases you would want to tweek the bandwidth and maybe other settings too.
